### PR TITLE
enhancement: Add filter options to the list policies method of the admin API

### LINKED ISSUE #297: Adding filtering capabilities to the list policies method of the admin API.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Feature description

We can enhance the list policies method by adding filtering. So, the implementation details are quite settled as we discussed before. The filtering can be done after all policies have been fetched from the store. With this, we won't need to maintain filtering logic across different store implementations. For the first iteration, the performance cost can be negligible.

So, how the filtering fields should look like? From the command line tool (`cerbosctl`) the filtering for `cerbosctl` can adopt a similar fashion with the [`kubectl`](https://kubernetes.io/docs/reference/kubectl/cheatsheet/#viewing-finding-resources). Taken this into consideration, as we discussed [here](https://github.com/cerbos/cerbos/pull/269#discussion_r706176817) we can leverage filtering with field selectors against policies.

1. We can add basic filtering (e.g. policy kinds).
	- the kind filter should accept multiple kinds
	- disabled field (this needs confirmation) 
2. We can add advanced filtering with the field selectors.
	- filter field with an exact match (the flag can be `--field-eq` for this type of filtering)
	- filter field with wildcard match (the flag can be `--field-match` for this type)
	- search arguments should be in the form of `<field>=<value>`. For instance, if someone is looking "foo" in the resource policy description they should provide `--field-match resourcePolicy.description=foo`. Or, if they want to search "foo" among all policy kinds `--field-match description=foo`.

The API method should take either a list of basic filters or a list of advanced filters as the parameter. One thing we should consider is that whether the filter kinds above are forced to be used separately.

### What would the ideal solution look like to you?

For demonstration purposes, I am going to exemplify this with the `cerbosctl` output. The ticket includes work on both API method and the cli tool.

Current Impl. when we list the policies
```
$ cerbosctl list
NAME              KIND           DEPENDENCIES      VERSION   
donald_duck       PRINCIPAL      -                 20210210  
leave_request     RESOURCE       my_derived_roles  20210210  
leave_request     RESOURCE                         staging   
my_derived_roles  DERIVED_ROLES  -                 -     
```

We should be able to filter by kind like this:
```
$ cerbosctl list --kind principal
NAME              KIND           DEPENDENCIES      VERSION   
donald_duck       PRINCIPAL      -                 20210210  
```

Or, by field match like this:
```
$ cerbosctl list --field-eq resourcePolicy.resource=leave_request
NAME              KIND           DEPENDENCIES      VERSION   
leave_request     RESOURCE       my_derived_roles  20210210  
leave_request     RESOURCE                         staging   
```

### Anything else?

I'd like to link #136 here as this is a subset of that ticket.

### PR DESCRIPTION
**Pull request by isdevx**
#### Description

Adds filtering with JSONPath. 

Fixes #297

#### Checklist 

- [x] The PR title has the correct prefix 
- [x] PR is linked to the corresponding issue
- [x] All commits are signed-off (`git commit -s ...`) to provide the [DCO](https://developercertificate.org/)

### CODE REVIEW

File: internal/svc/admin_svc.go [hunk 1]
```diff
@@ -133,9 +134,50 @@ func (cas *CerbosAdminService) ListPolicies(ctx context.Context, req *requestv1.
 		return nil, status.Error(codes.Internal, fmt.Sprintf("could not get policies: %s", err.Error()))
 	}
 
-	policies := make([]*policyv1.Policy, 0, len(units))
-	for _, pl := range units {
-		policies = append(policies, pl.Policy)
+	pMap := make(map[int]*policyv1.Policy)
```
[the remark this answers is not in the document]
Comment (isdevx) [reply internal/svc/admin_svc.go:137]: Nice catch 👍

File: internal/svc/admin_svc.go [hunk 2]
```diff
@@ -133,9 +134,50 @@ func (cas *CerbosAdminService) ListPolicies(ctx context.Context, req *requestv1.
...
+			value := getStringValue(val)
+
+			switch filter.Type {
+			case requestv1.ListPoliciesRequest_MATCH_TYPE_EXACT:
+				if value == filter.Value {
+					continue
+				}
+				delete(pMap, i)
+			case requestv1.ListPoliciesRequest_MATCH_TYPE_WILDCARD:
+				if strings.Contains(value, filter.Value) {
```
[the remark this answers is not in the document]
Comment (isdevx) [reply internal/svc/admin_svc.go:168]: Sounds cool, why not :) We can accept regex for `--field-match`.

File: docs/modules/cli/pages/ctl.adoc [hunk 3]
```diff
@@ -121,3 +123,29 @@ Use the arrow keys (or Vim keys kbd:[h], kbd:[j], kbd:[k], kbd:[l]) to scroll ho
 ----
 cerbosctl decisions --tail=20
 ----
+
+
+== `list`
+
+This command lists the policies stored in the Cerbos server. It accepts the filters below to enable filtering the list. The filters use JSONPath expressions to query fields.
```
Comment (charithe) [inline docs/modules/cli/pages/ctl.adoc:130]:
```suggestion
This command lists the policies available in the configured policy repository. The list can be filtered by specifying one or more pairs of field selectors and match values as command line arguments. Field selectors are expected to be valid JSONPath expressions describing policy fields.
```

File: docs/modules/cli/pages/ctl.adoc [hunk 4]
```diff
@@ -121,3 +123,29 @@ Use the arrow keys (or Vim keys kbd:[h], kbd:[j], kbd:[k], kbd:[l]) to scroll ho
...
+See link:https://tools.ietf.org/id/draft-goessner-dispatch-jsonpath-00.html#table-1[specification] for detailed information about JSONPath.
```
Comment (charithe) [inline docs/modules/cli/pages/ctl.adoc:131]:
```suggestion
See link:https://tools.ietf.org/id/draft-goessner-dispatch-jsonpath-00.html#table-1[specification] for more information about JSONPath.
```

File: docs/modules/cli/pages/ctl.adoc [hunk 5]
```diff
@@ -121,3 +123,29 @@ Use the arrow keys (or Vim keys kbd:[h], kbd:[j], kbd:[k], kbd:[l]) to scroll ho
...
+--field-eq:: Filter a field with an exact match
++ 
+- `--field-eq="$.resourcePolicy.version=20210210"`: Will list resource policies with version 20210210.
+- `--field-eq="$.*.version=20210210"`: Will list any type of policies with version 20210210.
+--field-match:: Filter a field with a regex match. 
++
+- `--field-match="$.resourcePolicy.resource=leave"`: Will list the resource policies with "leave" contained in the resource field.
+- `--field-match="$.resourcePolicy.version=\d"`: Will list resource policies with a version includes a numeric value.
+
+****
```
Comment (charithe) [inline docs/modules/cli/pages/ctl.adoc:146]: Let's keep the example style consistent with the rest of the documentation.
 
```suggestion
Use `--field-eq` to perform an exact match and `--field-match` to perform a regular expression match.

.List resource policies where the `version` field is exactly `20210210`
----
cerbosctl list --field-eq='$.resourcePolicy.version=20210210'
----

.List any policy where `version` is exactly `20210210`
----
cerbosctl list --field-eq='$.*.version=20210210'
----

.List resource policies where the `resource` field contains `leave`
----
cerbosctl list --field-match='$.resourcePolicy.resource=leave'
----

.List resource policies with a numeric version
----
cerbosctl list --field-match='$.resourcePolicy.version=\d+'
----
```

File: docs/modules/cli/pages/ctl.adoc [hunk 6]
```diff
@@ -121,3 +123,29 @@ Use the arrow keys (or Vim keys kbd:[h], kbd:[j], kbd:[k], kbd:[l]) to scroll ho
...
+
+
+== `version`
+
+This command displays the detailed version information of the cerbosctl and the Cerbos server it connected.
```
Comment (charithe) [inline docs/modules/cli/pages/ctl.adoc:151]: I don't think we need this section here.

```suggestion

```

**Review by charithe — COMMENTED** [review]
LGTM 🚀 

Just left a few suggestions below to keep the documentation consistent with the overall tone and style.

**Comment by isdevx** [comment]
Thanks for the suggestions @charithe! Applied those and updated the PR.

**Review by charithe — APPROVED** [review]
Nice one 👍🏽

**Pull request merged**